src/server.rs

```diff
@@ -69,6 +71,69 @@ fn replace_string_in_range(
     contents
 }
 
+fn function_defines(
+    name: &str,
+    params: &Vec<FunctionParameter>,
+) -> bool {
+    params.iter().any(|param| param.key.name == name)
+}
+
+fn is_scope(name: &str, n: Rc<walk::Node<'_>>) -> bool {
+    let mut dvisitor = DefinitionFinderVisitor::new(name.to_string());
+    walk::walk(&mut dvisitor, n.clone());
+    let state = dvisitor.state.borrow();
+
+    state.node.is_some()
+}
+
+fn find_references(
+    uri: lsp::Url,
+    result: NodeFinderResult,
+) -> Vec<lsp::Location> {
+    let mut locations: Vec<lsp::Location> = vec![];
+
+    if let Some(node) = result.node {
+        let name = match node.as_ref() {
+            walk::Node::Identifier(ident) => &ident.name[..],
+            walk::Node::IdentifierExpr(ident) => &ident.name[..],
+            _ => return locations,
+        };
+
+        let mut path_iter = result.path.iter().rev();
+        let scope: Option<Rc<walk::Node>> =
+            path_iter.find_map(|n| match n.as_ref() {
+                walk::Node::FunctionExpr(f)
+                    if function_defines(name, &f.params) =>
+                {
+                    Some(n.to_owned())
+                }
+                walk::Node::Package(_) | walk::Node::File(_)
+                    if is_scope(name, n.clone()) =>
+                {
+                    Some(n.to_owned())
+                }
+                _ => None,
+            });
+
+        if let Some(scope) = scope {
+            let mut visitor =
+                IdentFinderVisitor::new(name.to_string());
+            walk::walk(&mut visitor, scope);
+
+            let state = visitor.state.borrow();
+            let identifiers = (*state).identifiers.clone();
+
+            for node in identifiers {
```

Comment:
I think a more rust way of doing this, and wouldn't require a clone, is to do this:

```
let new_locations = (*state).identifiers.iter().map(|node| map_node_to_location(uri.clone(),  node.clone()))
locations.append(new_locations);
```

I'm not really sure what `borrow()` is so I'm not sure if the rest can also be changed if we use a reference here instead. I don't think you need `let mut locations` at the beginning of this function because it seems like you can just return it here.